clippy_lints/src/methods/case_sensitive_file_extension_comparisons.rs

```diff
@@ -28,13 +31,46 @@ pub(super) fn check<'tcx>(
         let recv_ty = cx.typeck_results().expr_ty(recv).peel_refs();
         if recv_ty.is_str() || is_type_lang_item(cx, recv_ty, LangItem::String);
         then {
-            span_lint_and_help(
+            span_lint_and_then(
                 cx,
                 CASE_SENSITIVE_FILE_EXTENSION_COMPARISONS,
-                call_span,
+                recv.span.to(call_span),
                 "case-sensitive file extension comparison",
-                None,
-                "consider using a case-insensitive comparison instead",
+                |diag| {
+                    diag.help("consider using a case-insensitive comparison instead");
+                    let mut recv_source = Sugg::hir(cx, recv, "").to_string();
+
+                    if is_type_lang_item(cx, recv_ty, LangItem::String) {
+                        recv_source = format!("&{recv_source}");
+                    }
+
+                    if recv_source.ends_with(".to_lowercase()") {
+                        diag.note("to_lowercase allocates memory, this can be avoided by using Path");
+                        recv_source = recv_source.strip_suffix(".to_lowercase()").unwrap().to_string();
+                    }
+
+                    if recv_source.ends_with(".to_uppercase()") {
+                        diag.note("to_uppercase allocates memory, this can be avoided by using Path");
+                        recv_source = recv_source.strip_suffix(".to_uppercase()").unwrap().to_string();
+                    }
```

Comment:
I would say detecting the use of `to_lower/uppercase` is good, but personally I think we should not lint when we detect that to be the case. Linting about a case sensitive operation when we know it's case insensitive seems a bit too confusing to me

Reply:
Ok, I don't understand what you are asking. My understanding of the original issue was that users were using to_lowercase as a work-around to the case-sensitive nature of `ends_with`. The whole point of this lint is to catch when users do that and let them know there is a case-insensitive way to check for extensions without the allocations of is_lowercase.

Are you suggesting that if we detect `is_lowercase` we should assume the user intended to allocate and convert to the lowercase and we should not lint in that case?

Reply:
The purpose of the lint is to detect case sensitive comparisons and suggest changing them to case insensitive ones, hence the message being `error: case-sensitive file extension comparison`. It happens to lint `s.to_lowercase().ends_with("...")` currently, but I agree with #10042 that this is a false positive

Yeah, we should not lint in that case

Reply:
Would you like me to change this to detect that case (`to_lowercase` at the end of the recv) and stop the lint?

Reply:
I pushed a change that skips the lint if to_lowercase or to_uppercase is the last method called.

Reply:
Great, thanks